### PR TITLE
Adding statistical tests to ISC and ISFC

### PR DESCRIPTION
**Pull request by cbaldassano**
Adding statistical testing through phase randomization, as requested by https://groups.google.com/forum/#!topic/brainiak/vZnOgGoV7q4 and as described in Simony et al. (2015, DOI: 10.1038/ncomms12141)

### CODE REVIEW

File: brainiak/isfc.py [hunk 1]
```diff
@@ -90,27 +143,135 @@ def isfc(D, collapse_subj=True):
...
+    if two_sided:
+        min_null_ecdf = ecdf(min_null)
+        p = 2 * np.minimum(1 - max_null_ecdf(ISFC), min_null_ecdf(ISFC))
+        p = np.minimum(p, 1)
+    else:
+        p = 1 - max_null_ecdf(ISFC)
+    return ISFC, p
+
+
+def phase_randomize(D):
```
Comment (mihaic) [inline brainiak/isfc.py]: This function is not really ISFC specific. I think it fits better in `utils/utils.py`.

File: brainiak/isfc.py [hunk 2]
```diff
@@ -90,27 +143,135 @@ def isfc(D, collapse_subj=True):
...
+    """Randomly shift signal phases
+
+    For each timecourse (from each voxel and each subject), computes its DFT
+    and then randomly shifts the phase of each frequency before inverting
+    back into the time domain. This yields timecourses with the same power
+    spectrum (and thus the same autocorrelation) as the original timecourses,
+    but will remove any meaningful temporal relationships between the
+    timecourses.
```
Comment (mihaic) [inline brainiak/isfc.py]: Do you have a paper reference by any chance?

File: brainiak/isfc.py [hunk 3]
```diff
@@ -90,27 +143,135 @@ def isfc(D, collapse_subj=True):
...
+    shift = np.random.rand(D.shape[0], len(pos_freq), D.shape[2]) * 2 * math.pi
+
+    # Shift pos and neg frequencies symmetrically, to keep signal real
+    F[:, pos_freq, :] *= np.exp(1j * shift)
+    F[:, neg_freq, :] *= np.exp(-1j * shift)
+
+    return np.real(ifft(F, axis=1))
+
+
+def ecdf(x):
```
Comment (mihaic) [inline brainiak/isfc.py]: `utils/utils.py`?

File: tests/isfc/test_isfc.py [hunk 4]
```diff
@@ -11,15 +11,20 @@ def test_ISC():
          [-0.62305959, -0.46660116, -0.50037994,  1.81083754,  0.23499509]]
     D[:, :, 1] = \
         [[-0.31484153, -0.47486988,  0.18966625, -0.12568572, -0.40535156],
-         [-0.78433298, -0.35875085, -0.6121344,  1.68267639,  0.28603493]]
+         [1.68267639, -0.78433298, -0.35875085, -0.6121344,  0.28603493]]
     D[:, :, 2] = \
         [[-0.26593192, -0.56914734,  0.28397317,  0.30276589, -0.42637472],
-         [-0.67598379, -0.51549055, -0.64196342,  1.60686666,  0.04127293]]
+         [0.04127293, -0.67598379, -0.51549055, -0.64196342,  1.60686666]]
```
[the remark this answers is not in the document]
Comment (cbaldassano) [reply tests/isfc/test_isfc.py]: Added a comment to clarify, I wanted to create a null ISC to make sure it was not significant.

File: tests/isfc/test_isfc.py [hunk 5]
```diff
@@ -35,13 +40,20 @@ def test_ISFC():
 
     assert D.shape == (4, 5, 2), "Loaded data has incorrect shape"
 
-    ISFC = brainiak.isfc.isfc(D)
+    np.random.seed(0)
```
[the remark this answers is not in the document]
Comment (mihaic) [reply tests/isfc/test_isfc.py]: Did you actually forget to remove the line? :)

File: brainiak/utils/utils.py [hunk 6]
```diff
@@ -667,3 +671,78 @@ def usable_cpu_count():
         except AttributeError:
             result = os.cpu_count()
     return result
+
+
+def phase_randomize(D, random_state=0):
```
Comment (mihaic) [inline brainiak/utils/utils.py]: Please add tests for the new functions. They can be as simple as you want.

**Pull request merged**